data migration from v2 metrics tables to v3 metrics tables and cleanu…

gsps1 opened the pull request:
…p v2 metrics table after migration is done

JIRA :  https://issues.cask.co/browse/CDAP-12348

gsps1 committed:
CDAP-12348 data migration from v2 metrics tables to v3 metrics tables and cleanup v2 metrics table after migration is done

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -271,7 +332,7 @@ private void persistMetrics(Deque<MetricValues> metricValues) throws Exception {
     metricStore.add(metricValues);
     metricsProcessedCount += metricValues.size();
-    PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric metric's timestamp: {}. " +
+      PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric metric's timestamp: {}. " +
```
unintended indent?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -194,6 +214,47 @@ protected void run() {
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-table-deleter"));
+
+        // scheduled to run every 2 hours
+        scheduledExecutorServiceDelete.scheduleAtFixedRate(tableDeletion, 1, 2, TimeUnit.HOURS);
```
why is the deletion in a separate thread? Can't the dataMigrator delete the table when it has migrated all data?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+   */
+  public boolean isOldMetricsDataAvailable() {
+    Scanner scanner = v2MetricsTable.scan(null, null, null);
+    Row row = scanner.next();
```
you should use a try (scanner =) { ... } finally { close }

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+   * @param maxRecordsToScan - limit on number of records scanned
+   */
+  public void transferData(int maxRecordsToScan) {
+    LOG.trace("Running Metrics Migrator Thread");
```
isn't this thread already running? You mean "running metrics migration for {} records", maxRecords?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+
+    while (((row = scanner.next()) != null) && (recordsScanned < maxRecordsToScan)) {
+      if (recordsScanned % 10 == 0) {
+        LOG.trace("Scanned {} records in Metrics Data Migration", recordsScanned);
```
every 10 records?

gsps1 replied:
yeah, trace log, will make it to print for every 100 records instead ? or should i remove it ?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+    Row row;
+    int recordsScanned = 0;
+
+    while (((row = scanner.next()) != null) && (recordsScanned < maxRecordsToScan)) {
```
check (recordsScanned < maxRecordsToScan) first; you can avoid the IO if that is false.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+        v2MetricsTable.delete(rowKey, deletes);
+      }
+
+      if (!gauges.isEmpty()) {
```
for the gauges, you need to prevent a conflict with the metrics processor. It might write a value to the v3 table (and delete the one in v2, or keep in in v2 if it is actually an increment) between the time you read the row and and the time you write the gauges to v3. Therefore you need to use checkAndPut, and retry if the v3 value has changed.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+        rowIncrements.put(rowKey, increments);
+        v3MetricsTable.increment(rowIncrements);
+      }
+      // delete entries from old table
```
misplaced comment?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+   */
+  public void transferData(int maxRecordsToScan) {
+    LOG.trace("Running Metrics Migrator Thread");
+    Scanner scanner = v2MetricsTable.scan(null, null, null);
```
here, too: use a try ... finally for making sure the scanner is closed.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+        TableId hBaseTableId =
+          tableUtil.createHTableId(new NamespaceId(tableId.getNamespace()), tableId.getTableName());
+        boolean doesExist  = tableUtil.tableExists(admin, hBaseTableId);
+        LOG.trace("Table {} exists ? : {}", hBaseTableId.getTableName(), doesExist);
```
remove " ? "

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -115,6 +131,32 @@ protected MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperti
+    try {
+      // metrics tables are in the system namespace
+      DatasetId tableId = NamespaceId.SYSTEM.dataset(tableName);
+      MetricsTable table = null;
```
I think you can just call getDataset() and it will return null if the table does not exist.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,206 @@
+        if (tableUtil.tableExists(admin, hBaseTableId)) {
+          LOG.trace("Found table {}, going to delete", hBaseTableId);
+          tableUtil.dropTable(ddlExecutor, hBaseTableId);
+          LOG.trace("Deleted table {}", hBaseTableId);
```
what happens if a query is currently reading that table? Do you need to change something in the read path to catch the exception and retry?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -194,6 +214,47 @@ protected void run() {
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-data-migrator"));
+
+        // 5-min initial delay and scheduled to run each min after that
+        scheduledExecutorService.scheduleAtFixedRate(dataMigrator, 5, 1, TimeUnit.MINUTES);
```
when do these scheduled executors terminate? Looks like they will keep triggering after all migration is done.

anew reviewed (changes requested):
please see my comments

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetricsTable.java:
```diff
@@ -84,6 +86,18 @@
   void delete(byte[] row, byte[][] columns);
 
   /**
+   * checks the expected value before performing a put
```
if you have @param, then it should actually describe what the parameter means. Also, start a Javadoc with an upper case letter, and end the sentence with a period?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -303,6 +303,19 @@ public void delete(byte[] row, byte[][] columns) {
+    try {
+      return hTable.checkAndPut(row, columnFamily, qualifier, expectedValue, put);
+    } catch (IOException e) {
+      throw new DataSetException("Put failed on table " + tableId, e);
```
CheckAndPut failed

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -227,31 +234,39 @@ protected void run() {
-        dataMigrator.run();
+        int maxRecordsToScan = 1000;
+        LinkedHashMap<Integer, Integer> resolutionToMaxRecordsMap = new LinkedHashMap<>();
+        // migrate aggregates resolution table without limit on max records scan
```
why?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -227,31 +234,39 @@ protected void run() {
-        // migrate 1hr resolution table first, then 1min resolution table, limit 1000 records to scan/transfer in a run.
         dataMigrator = new DataMigrator(metricDatasetFactory, cConfiguration, hConf, hBaseTableUtil,
-                                        ImmutableList.of(3600, 60), 1000);
+                                                     resolutionToMaxRecordsMap);
```
please fix indentation

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -227,31 +234,39 @@ protected void run() {
-        // 5-min initial delay and scheduled to run each min after that
-        scheduledExecutorService.scheduleAtFixedRate(dataMigrator, 5, 1, TimeUnit.MINUTES);
-        LOG.info("Scheduled metrics migration thread for resolution 3600, 60 tables");
+          // 5-min initial delay and scheduled to run each min after that
```
why 5 minutes delay?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -332,7 +347,7 @@ private void persistMetrics(Deque<MetricValues> metricValues) throws Exception {
     metricStore.add(metricValues);
     metricsProcessedCount += metricValues.size();
-      PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric metric's timestamp: {}. " +
+    PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric metric's timestamp: {}. " +
```
duplicate word metric

anew replied:
actually twice in this message

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -380,6 +395,19 @@ public void run() {
+          // todo uncomment and run only every 100 times
+          /*if (instanceId == 0 && topicIdMetaKey.getTopicId().equals(metricsTopics.get(0))) {
+            if (dataMigrator != null && dataMigratorExecutor != null && !dataMigratorExecutor.isShutdown()
+              && dataMigrator.isMigrationComplete()) {
```
I am not sure about this, but can the migrator itself shut down the executor when it is done with the migration?

anew replied:
It seems that that would be ok: https://stackoverflow.com/questions/29698934/shutting-down-an-executorservice-from-running-task

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -380,6 +395,19 @@ public void run() {
+              dataMigratorExecutor.shutdown();
+            }
+            if (tableDeleter != null && tableDeleterExecutor != null && !tableDeleterExecutor.isShutdown()
+              && tableDeleter.allTablesDeleted()) {
```
I really don't think you want to call this every 100 seconds.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
+      Row row;
+      int recordsScanned = 0;
+      while ((recordsScanned < maxRecordsToScan) && ((row = scanner.next()) != null)) {
+        if (recordsScanned % 10 == 0) {
```
that's gonna create a lot of logs... this should be trace

anew replied:
oh, ignore my comment, I initially did not see this commit.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
+          // column is timestamp, do a get on the new table
+          byte[] value = v3MetricsTable.get(rowKey, entry.getKey());
+          if (value == null) {
+            LOG.debug("Trying checkAndPut");
```
trace

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
+          if (value == null) {
+            LOG.debug("Trying checkAndPut");
+            if (v3MetricsTable.checkAndPut(rowKey, entry.getKey(), new byte[0], entry.getValue())) {
+              LOG.debug("checkAndPut Successful");
```
trace

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
+          byte[] value = v3MetricsTable.get(rowKey, entry.getKey());
+          if (value == null) {
+            LOG.debug("Trying checkAndPut");
+            if (v3MetricsTable.checkAndPut(rowKey, entry.getKey(), new byte[0], entry.getValue())) {
```
is a checkAndPut of an empty byte array equivalent to a delete? I guess that belongs into the Javadocs for checkAndPut...

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
+              LOG.debug("checkAndPut Successful");
+              gaugeDeletes.add(entry.getKey());
+            } else {
+              LOG.debug("checkAndPut failed");
```
trace.

anew replied:
What is the idea here? If the checkAndPut fails, the column remains unmodified and will be migrated in the next iteration? It would be good to explain the approach in a comment

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -82,67 +84,69 @@ public boolean isOldMetricsDataAvailable() {
-      if (!increments.isEmpty()) {
-        rowIncrements.put(rowKey, increments);
-        v3MetricsTable.increment(rowIncrements);
+        // break if we have exceeded max records to scan
```
this comment does not seem right... I don't see the break

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -167,7 +171,7 @@ public boolean v2MetricsTableExists(HBaseTableUtil tableUtil, int resolution) {
           tableUtil.createHTableId(new NamespaceId(tableId.getNamespace()), tableId.getTableName());
         boolean doesExist  = tableUtil.tableExists(admin, hBaseTableId);
-        LOG.trace("Table {} exists ? : {}", hBaseTableId.getTableName(), doesExist);
+        LOG.debug("Table {} exists : {}", hBaseTableId.getTableName(), doesExist);
```
is this called often? Perhaps you only want to log that if it exists, because that means the migration is not done yet?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -186,15 +190,15 @@ public boolean v2MetricsTableExists(HBaseTableUtil tableUtil, int resolution) {
                                       HBaseTableUtil tableUtil, int resolution) {
     TableId tableId = getV2MetricsTableName(resolution);
-    LOG.trace("Looking to delete table {}", tableId);
+    LOG.debug("Looking to delete table {}", tableId);
```
too much logging here

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -43,12 +54,51 @@
-  public CombinedHBaseMetricsTable(MetricsTable v2HBaseTable, MetricsTable v3HBaseTable) {
+  private boolean v2TableDeleted;
+
+  public CombinedHBaseMetricsTable(final MetricsTable v2HBaseTable, MetricsTable v3HBaseTable,
```
this needs Javadocs.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -43,12 +54,51 @@
+  }
+
+   @VisibleForTesting
+   CombinedHBaseMetricsTable(MetricsTable v2HBaseTable, MetricsTable v3HBaseTable,
```
Javadocs.... what is the predicate for?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -59,7 +109,9 @@ public CombinedHBaseMetricsTable(MetricsTable v2HBaseTable, MetricsTable v3HBase
   public void put(SortedMap<byte[], ? extends SortedMap<byte[], Long>> updates) {
-    deleteColumns(updates);
+    if (tableExists.apply(resolution)) {
+      deleteColumns(updates);
```
you check whether the table exists before every single put? How is that going to perform?

anew replied:
also, this does not protect you from exceptions, because the table may be disabled or get deleted between the check and the delete. Instead, you need to catch IOException from the v2Table, and if it does not exist any longer, remember that (for example, by setting v2Table to null) and not attempt to write to it again.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -116,14 +170,47 @@ public boolean checkAndPut(byte[] row, byte[] column, byte[] expectedValue, byte
     v3HBaseTable.close();
-    v2HBaseTable.close();
+    if (tableExists.apply(resolution)) {
+      v2HBaseTable.close();
```
you must close this in any case.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -253,6 +253,8 @@ protected void run() {
           dataMigratorExecutor.scheduleAtFixedRate(dataMigrator, 5, 1, TimeUnit.MINUTES);
           LOG.info("Scheduled metrics migration thread for resolution INT_MAX, 3600, 60 tables");
+        } else {
+          LOG.info("All Metrics data have been migrated, not scheduling migration thread");
```
I think this should be at most debug, if it should be logged at all. To the user it is only of interest if the migratuon was started, the first time after upgrade. Other than that, this message is only confusing

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -266,6 +268,8 @@ protected void run() {
           LOG.info("Scheduled metrics deletion thread for 1, 60, 3600, INT_MAX resolution tables, " +
                      "tables will only be deleted after data migration is completed for them");
+        } else {
+          LOG.info("All Metrics tables have been deleted, not scheduling deletion thread");
```
same here

anew commented:
@gsps1 please see my comments. It looks like you made changes while I was reviewing, so my comments don't show properly, please read and address all my comments.

gsps1 commented:
Hi @anew thanks for the review. I have addressed the comments.
1) https://github.com/caskdata/cdap/pull/9399/commits/7b03b56ccc7685491508f0ceea807a792148569b - realized we already have swap which does checkAndPut, so removed checkAndPut, should have found that earlier. 
2) https://github.com/caskdata/cdap/pull/9399/commits/3712e3e46ad3fe9d56bebeac4e35c87b0dbd4271 - changes in combinedMetricsTable to handle deleted v2 table from exception, removed predicate usage, as it wasn't necessary with this change.

3) https://github.com/caskdata/cdap/pull/9399/commits/b5c9fccbdaf0dfc50c84b439a80c3e70abecba85
- made changes to stop the executor service from migrator, deleter classes once they are complete, per your suggestion. Also added retry logic for getting the metrics tables in the constructor. There are some minor code duplication, i will remove them if the approach is correct, wanted to get your opinion on it before.

Please take a look again, thanks!

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -158,25 +136,41 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
     v3HBaseTable.delete(row, columns);
-    if (tableExists.apply(resolution)) {
+    try {
       v2HBaseTable.delete(row, columns);
```
this will always throw NPE after v2Table was deleted. You need to check whether it is null.

gsps1 replied:
my bad, missed checking at this place. have fixed it now.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -24,38 +25,72 @@
+
 
   // Data migrator thread will be created for each resolution tables, so they can be scheduled appropriately.
   public DataMigrator(MetricDatasetFactory metricDatasetFactory, CConfiguration cConf, Configuration hConf,
```
public methods/constructors need Javadoc

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -67,38 +102,44 @@ public boolean isMigrationComplete() {
       MetricsTableMigration metricsTableMigration = migrationOrder.get(resolution);
-      if (metricsTableMigration.isOldMetricsDataAvailable()) {
+      if (metricsTableMigration.v2MetricsTableExists(hBaseTableUtil, resolution) &&
+        metricsTableMigration.isOldMetricsDataAvailable()) {
```
doesn't the second condition imply the first one?

gsps1 replied:
if the table is not available, it would be better to skip doing scan to check if metrics data is available as that will throw an exception, so decided to do that only if table exists and remove the entry if the table no longer exists.
i also noticed that i wasn't handling possible exception say the table gets deleted while doing this check or any other exception during data transfer, i have added catch block to handle that currently.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -351,7 +329,7 @@ private void persistMetrics(Deque<MetricValues> metricValues) throws Exception {
     metricStore.add(metricValues);
     metricsProcessedCount += metricValues.size();
-    PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric metric's timestamp: {}. " +
+    PROGRESS_LOG.debug("{} metrics metrics persisted. Last metric's timestamp: {}. " +
```
stlll a double "metrics".

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,127 @@
+        }
+        LOG.warn("Cannot access v2 metricsTable, will retry in 1 sec.");
+        try {
+          TimeUnit.SECONDS.sleep(1);
```
why retry forever? What if the table has some other issue? Also, why not sleep longer?

gsps1 replied:
updated the logic to perform table get in scheduled run and not during initialization on this commit. https://github.com/caskdata/cdap/pull/9399/commits/0cf263b5c766af5e7e050680fb610160ea0b1a82

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,127 @@
+        LOG.warn("Cannot access v3 metricsTable, will retry in 1 sec.");
+        try {
+          TimeUnit.SECONDS.sleep(1);
+        } catch (InterruptedException ie) {
```
same here

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -19,80 +19,60 @@
+    try {
+      v2table = metricDatasetFactory.getV2MetricsTable(resolution);
+    } catch (Exception e) {
+      LOG.debug("Cannot access v2 metricsTable due to exception", e);
```
maybe this should be a warning or at least info. It prevents migration, so it must be surfaced. Also, if the table does not exist, don't you want to handle that differently?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -19,80 +19,60 @@
+    try {
+      v3table = metricDatasetFactory.getV3MetricsTable(resolution);
+    } catch (Exception e) {
+      LOG.debug("Cannot access v3 metricsTable due to exception", e);
```
same here

anew reviewed (commented):
please see my comments.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetricsTable.java:
```diff
@@ -16,8 +16,10 @@
 
 package co.cask.cdap.data2.dataset2.lib.table;
 
+import co.cask.cdap.api.annotation.Beta;
```
unused imports? Beta and Put

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +136,68 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+    }
+  }
+
+  private void handleTableException(Exception e) {
```
better name this method handleV2TableException? To make it clear what it is used for? Or call it closeV2TableIfDeleted(). Also - a comment never harms: what is the purpose of this method.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +136,68 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+        Closeables.closeQuietly(v2HBaseTable);
+        v2HBaseTable = null;
+      }
+    }
```
otherwise, I believe you still want to rethrow the original exception?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +136,68 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+    }
+  }
+
+  private TableId getV2MetricsTableName(int resolution) {
```
does resolution have to be a parameter? Or is it always the same as the instance variable of the same name?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +136,68 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
     v3HBaseTable.close();
-    v2HBaseTable.close();
+    if (v2HBaseTable != null) {
+      v2HBaseTable.close();
```
can close() ever throw an exception? Better use Closeables.closeQuietly()

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -105,12 +105,16 @@ public Row next() {
   public void close() {
     v3TableScanner.close();
-    v2TableScanner.close();
+    if (v2TableScanner != null) {
```
here, too: Better use Closeables.closeQuietly() for both scanners

gsps1 replied:
these scanner's extend AutoCloseable (java.lang) and they override the close to not throw exception..

anew replied:
then use a try {...} finally

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -105,12 +105,16 @@ public Row next() {
     Row resultRow = v2TableNextRow;
-    v2TableNextRow = v2TableScanner.next();
+    if (v2TableScanner != null) {
+      v2TableNextRow = v2TableScanner.next();
```
if the table was deleted, this will throw an exception. You need catch and check whether the table was deleted, similar other combined table operations. If it was deleted, set v2scanner to null and the scan will continue with v3 only.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+   * Data migrator to schedule data migration for metrics tables from v2 to v3 metrics tables.
+   * @param metricDatasetFactory metric dataset factory
+   * @param cConf cConfiguration
+   * @param hConf hConfiguratioin
```
type. You could at least document this as HBase configuration.  Or remove the @param annotations for parameters that are self explanatory. Repeating the parameter name as its description is pointless.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+      scheduledExecutorService.scheduleAtFixedRate(new MigrationTask(), 5, 1, TimeUnit.MINUTES);
+      LOG.info("Scheduled metrics migration thread for resolution INT_MAX, 3600, 60 tables");
+    } else {
+      scheduledExecutorService.shutdown();
```
you can also set it to null let it be GC'd.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+  private boolean isMigrationComplete() {
+    Iterator<Integer> resolutions = maxRecordsToScanResolutionMap.keySet().iterator();
+
+    if (!resolutions.hasNext()) {
```
this is not really needed because in this case, the while loop will not execute.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+      return true;
+    }
+
+    // when metrics.processor restarts and initializes DataMigrator,
```
seems that this comment belongs into the Javadoc for this method

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+  }
+
+  /**
+   * returns true when data migration is complete and
```
please fix all your Javadocs. Sentences start with upper case and end with a period. There is a blank line before the @param section. Please look at other Javadocs if you are in doubt about what a Javadoc should look like.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+      int resolution = resolutions.next();
+      if (migrationTableUtility.v2MetricsTableExists(resolution)) {
+        try (MetricsTable v3MetricsTable = migrationTableUtility.getV3MetricsTable(resolution);
+             MetricsTable v2MetricsTable = migrationTableUtility.getV2MetricsTable(resolution)) {
```
I guess the name of the method (get...Table) suggests that this will fail if there is an exception. It would be clearer if the method name indicates that it returns null if there is an exception. 

Also, is this true for all possible exceptions, that you want to ignore them and try again next cycle?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+            new MetricsTableMigration(v2MetricsTable, v3MetricsTable, cConf);
+          // if metrics data is available in old table, return false
+          if (metricsTableMigration.isOldMetricsDataAvailable()) {
+            return false;
```
I am confused by this whole logic. Isn't this a lot easier: 
- schedule migration whenever any v2 table exists
- migration finds out whether there is data
  - if so, it migrates the data and then deletes the v2 table
  - if not, it deletes the table

I don't see much value in trying to determine here whether a table needs migration or deletion, and then have two different tasks for migration and deletion.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+      // if the table has been deleted for that resolution or if there are no more data available in the table
+      while (resolutions.hasNext()) {
+        int resolution = resolutions.next();
+        if (migrationTableUtility.v2MetricsTableExists(resolution)) {
```
this code seems like a repetition of the code in isMigrationComplete().

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -218,6 +236,43 @@ protected void run() {
     }
 
+    if (instanceId == 0) {
+      // todo iterate migration resolution order list and schedule execution based on data availability.
```
is this still a todo?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -218,6 +236,43 @@ protected void run() {
+    if (instanceId == 0) {
+      // todo iterate migration resolution order list and schedule execution based on data availability.
+
+      // transfer aggregates metrics data at the beginning, in a single run to avoid conflicts with new metrics written
```
is this comment still accurate? I think you changed the logic since then

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -218,6 +236,43 @@ protected void run() {
+      // to the total unique metrics in the system.
+      // we will schedule the transfer of metrics data for time resolution tables
+      // in a single thread executor with limit on records to transfer.
+      // since the smaller resolution tables have timestamp information in the rowkey, they wont have conflict for most
```
are you saying that results of the migration can be incorrect due to conflicts, and it does not matter because it is only a small chance? really?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -218,6 +236,43 @@ protected void run() {
+      if (!skipMigration) {
+        int maxRecordsToScan = 1000;
+        LinkedHashMap<Integer, Integer> resolutionToMaxRecordsMap = new LinkedHashMap<>();
+        // migrate aggregates resolution table without limit on max records scan
```
still not clear to me what's the advantage of having no limit?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableDeleter.java:
```diff
@@ -0,0 +1,129 @@
+                 "tables will only be deleted after data migration is completed for them");
+    } else {
+      LOG.debug("Not Scheduling Deletion thread, all tables have already been deleted");
+      scheduledExecutorService.shutdown();
```
and set it to null.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -218,6 +236,43 @@ protected void run() {
+        resolutionToMaxRecordsMap.put(60, maxRecordsToScan);
+
+        ScheduledExecutorService dataMigratorExecutor =
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-data-migrator"));
```
who will shutdown these executors if metrics processor is shutdown? (before the migration is complete)

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,177 @@
+                resolutions.remove();
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception while performing metrics data transfer", e);
```
what if it is an interruptedException? That would happen when metrics processor shuts down and calls shutdownNow() for the executor. (does it? - it should)

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableDeleter.java:
```diff
@@ -0,0 +1,129 @@
+        return;
+      }
+
+      for (int resolution : resolutionsToDelete) {
```
This has lots of code duplication with the data migrator.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,143 @@
+            // if it was a gauge, the processing would have deleted the entry from the old table
+            // if it was a incrment, we would perform increment in the next iteration of the migration run
+            // so its safe to skip delete on checkAndPut failure
+            if (v3MetricsTable.swap(rowKey, entry.getKey(), null, entry.getValue())) {
```
it would be good idea to trace log if the swap fails. That would help us debug if the migration gets stuck.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,143 @@
+    return deletes;
+  }
+
+  private TableId getV2MetricsTableName(int resolution) {
```
this is at least the third time I see this code in a helper method.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+    this.tableUtil = tableUtil;
+  }
+
+  // we try to get v2 metrics table; if there is any exception while getting the table we return null
```
make this a Javadoc and please name the method in a way that indicates what it does. (attempts to get and returns null if that throws an exception)

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+    MetricsTable v2table = null;
+    try {
+      v2table = metricDatasetFactory.getV2MetricsTable(resolution);
+    } catch (Exception e) {
```
any exception? What if it is an OutOfMemory? Or IllegalArgumentException? Isn't there a (set of) expected exception(s) if the system is still starting up?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+    MetricsTable v3table = null;
+    try {
+      v3table = metricDatasetFactory.getV3MetricsTable(resolution);
+    } catch (Exception e) {
```
same here

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+   * check if v2 metrics table exists for the resolution
+   * @return true if table exists; false otherwise
+   */
+  boolean v2MetricsTableExists(int resolution) {
```
where else does this method exist? Looks like duplicate code to me.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+  boolean deleteV2MetricsTable(int resolution) {
+    TableId tableId = getV2MetricsTableName(resolution);
+    try {
+      try (HBaseDDLExecutor ddlExecutor = hBaseDDLExecutorFactory.get(); HBaseAdmin admin = new HBaseAdmin(hConf)) {
```
You should use the dataset framework to delete the table. Otherwise it will still exist as a dataset.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java:
```diff
@@ -0,0 +1,129 @@
+    return false;
+  }
+
+  private TableId getV2MetricsTableName(int resolution) {
```
more code duplication

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java:
```diff
@@ -73,11 +80,25 @@ public EntityTable get() {
 
+  @Override
+  public MetricsTable getV3MetricsTable(int resolution) {
+    String v3TableName = cConf.get(Constants.Metrics.METRICS_TABLE_PREFIX,
```
and more duplication!

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +136,68 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+      try (HBaseAdmin admin = new HBaseAdmin(hConf)) {
+        TableId hBaseTableId =
+          hBaseTableUtil.createHTableId(new NamespaceId(tableId.getNamespace()), tableId.getTableName());
+        return hBaseTableUtil.tableExists(admin, hBaseTableId);
```
why not use the dataset framework to do this?

anew reviewed (changes requested):
still lots of comments:
- I see too much copied/duplicate code
- the logic appears unnecessarily complex
- why bypass the dataset framework for table admin?
- be careful to make sure that all threads are shutdown and metrics processor cannot hang on shutdown. 
- please see all my comments

anew commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2081,6 +2081,15 @@
+    <name>metrics.table.migration.sleep.millis</name>
+    <value>10</value>
+    <description>
+      Delay in milliseconds between record transfers from v2 metrics table to v3 metrics table,
```
Delay in milliseconds after migration of each row from pre-4.3 to post 4.3 metrics tables. This is used to throttle the metrics migration and reduce its the load on the metrics system.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +134,74 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+          Closeables.closeQuietly(v2HBaseTable);
+          v2HBaseTable = null;
+        } else {
+          Throwables.propagate(e);
```
as a general style, always throw the result of propagate()

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +134,73 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+   * @param e
+   */
+  private void handleV2TableException(Exception e) {
+    if (e instanceof IOException) {
```
I think this can only be DataSetException or RuntimeException (from propagate()). In both cases, you need to check the cause for IOException.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +134,73 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+        }
+      } catch (DatasetManagementException dme) {
+        LOG.error("Exception while checking dataset exists", dme);
+        Throwables.propagate(e);
```
always throw the result of propagate()

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -106,19 +134,73 @@ public long incrementAndGet(byte[] row, byte[] column, long delta) {
+      }
+    }
+    // throw if any other exception or if the table exists
+    Throwables.propagate(e);
```
here, too.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -105,15 +117,48 @@ public Row next() {
+    try {
+      return v2TableScanner.next();
+    } catch (Exception e) {
+      if (e instanceof IOException) {
```
see the above discussion on handleTableExcpetion?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,91 @@
+  @Override
+  public void run() {
+    // iterate through resolutions, if datasetFramework has v2metricsTable then try to get V2 table with a retry
+    // once you get v2 table, get v3 table and run migration
```
nice comment, please add that you also delete the v2 table

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -0,0 +1,91 @@
+            metricsTableMigration.transferData(sleepMillisBetweenTransfer);
+            LOG.info("Metrics table data migration is complete for resolution {}", resolution);
+            // now transfer is complete; its safe to delete the v2 metrics table
+            MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, v3MetricsDatasetId);
```
don't delete v3

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -237,6 +252,34 @@ protected void run() {
+        int migrationSleepMillis =
+          Integer.valueOf(cConfiguration.get(Constants.Metrics.METRICS_MIGRATION_SLEEP_MILLIS));
+        metricsDataMigrator = new DataMigrator(datasetFramework, resolutions,
+                                                     v2TableNamePrefix, v3TableNamePrefix, migrationSleepMillis);
```
indent is off

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -245,6 +288,14 @@ protected void run() {
       }
     }
+
+    try {
```
maybe safer not to join, but set it to daemon and just exit. Otherwise the join my make the process hang on shutdown

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java:
```diff
@@ -257,6 +308,13 @@ protected void triggerShutdown() {
       thread.interrupt();
     }
+    if (metricsTableDeleterExecutor != null) {
+      metricsTableDeleterExecutor.shutdown();
```
consider using shutdownNow()

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableDeleter.java:
```diff
@@ -0,0 +1,52 @@
+        MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, datasetId);
+      }
+    } catch (Exception e) {
+      LOG.debug("Exception while performing dataset operation", e);
```
because we don't recover automatically from this, it should be an error

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,129 @@
+   */
+  public void transferData(int sleepMillisBetweenRecords) throws InterruptedException {
+    try (Scanner scanner = v2MetricsTable.scan(null, null, null)) {
+      LOG.trace("Starting Metrics Data Migration with {} ms sleep between records",
```
make it info

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,129 @@
+          if (value == null) {
+            // we perform checkAndPut for the new value, if the value was updated by the new upcoming metrics,
+            // if it was a gauge, the processing would have deleted the entry from the old table
+            // if it was a incrment, we would perform increment in the next iteration of the migration run
```
typo in incrment

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,129 @@
+        NavigableMap<byte[], Long> increments = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+        for (Map.Entry<byte[], byte[]> entry : columns.entrySet()) {
+          // column is timestamp, do a get on the new table
```
here you can check whether stop was requested and break the loop

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,129 @@
+                sleepMillisBetweenRecords);
+      Row row;
+      int recordsScanned = 0;
+      while (((row = scanner.next()) != null)) {
```
and not stop requested

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -0,0 +1,129 @@
+        recordsScanned++;
+        TimeUnit.MILLISECONDS.sleep(sleepMillisBetweenRecords);
+      }
+      LOG.debug("Migrated {} records from the metrics table {}", recordsScanned, v2MetricsTable);
```
this can be info

anew reviewed (commented):
please see comments :)

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableHelper.java:
```diff
@@ -31,10 +31,16 @@
  */
 final class MigrationTableHelper {
 
+  private static volatile boolean stopping = false;
```
@anew Am not too sure about this approach, will appreciate your suggestion if this is okay

anew replied:
I am not sure. Static volatile... might work. But returning null? Do the callers expect the null?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -151,20 +152,22 @@ public void delete(byte[] row, byte[][] columns) {
-          v2HBaseTable = null;
-          return;
+    if (e instanceof DataSetException || e instanceof RuntimeException) {
+      if (e.getCause() != null && e.getCause() instanceof IOException) {
```
no need to check for null, instanceof handles that

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -151,20 +152,22 @@ public void delete(byte[] row, byte[][] columns) {
+            return;
+          }
+        } catch (DatasetManagementException dme) {
+          LOG.error("Exception while checking dataset exists", dme);
```
this is used for various purposes, not only existence check. You may need to pass in the operation to be able to issue an accurate message here.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -184,9 +187,12 @@ public Scanner scan(@Nullable byte[] start, @Nullable byte[] stop, @Nullable Fuz
-    if (v2HBaseTable != null) {
-      Closeables.closeQuietly(v2HBaseTable);
+    try {
+      v3HBaseTable.close();
```
why can't you use closeQuietly()? You are using it for v3 table

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -143,16 +147,18 @@ private Row getV2TableScannerNextRow() {
-            v2TableScanner = null;
-            return null;
+      if (e instanceof DataSetException || e instanceof RuntimeException) {
+        if (e.getCause() != null && e.getCause() instanceof IOException) {
```
don't need to check for null

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -143,16 +147,18 @@ private Row getV2TableScannerNextRow() {
+              return null;
+            }
+          } catch (DatasetManagementException dme) {
+            LOG.error("Exception while checking dataset exists", dme);
```
as above, the message is wrong

gsps1 replied:
in this case, this exception is when we are checking if dataset instance exists.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -44,17 +46,29 @@
    */
   public DataMigrator(DatasetFramework datasetFramework, List<Integer> resolutions,
                       String v2TableNamePrefix, String v3TableNamePrefix, int sleepMillisBetweenTransfer) {
+    super(String.format("DataMigratorThread"));
```
use "metrics-migration". "Data" is too generic

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -44,17 +46,29 @@
+    metricsTableMigration = null;
+  }
+
+  public void setStopping() {
```
call this requestStop() or simply stop()

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -69,8 +83,11 @@ public void run() {
-            // now transfer is complete; its safe to delete the v2 metrics table
-            MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, v3MetricsDatasetId);
+            // don't delete if we are stopping
+            if (!stopping) {
```
better:
```
if (stopping) {
  break;
}
```

that will also exit the loop.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableDeleter.java:
```diff
@@ -44,7 +44,7 @@ public void run() {
       }
     } catch (Exception e) {
-      LOG.debug("Exception while performing dataset operation", e);
+      LOG.error("Exception while performing dataset operation", e);
```
please state what operation: deleting table "...".

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -45,9 +40,19 @@
   private final MetricsTable v2MetricsTable;
   private final MetricsTable v3MetricsTable;
 
+  private boolean stopping;
```
it must be volatle

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java:
```diff
@@ -58,11 +63,11 @@ public MetricsTableMigration(MetricsTable v2MetricsTable, MetricsTable v3Metrics
       Row row;
       int recordsScanned = 0;
-      while (((row = scanner.next()) != null)) {
+      while (((row = scanner.next()) != null) && !stopping) {
```
check stopping first, it is cheaper.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableHelper.java:
```diff
@@ -50,11 +56,16 @@ static MetricsTable getDatasetWithRetry(final DatasetFramework datasetFramework,
+        if (!stopping) {
+          return datasetFramework.getDataset(v2MetricsTableId, Collections.<String, String>emptyMap(), null);
+        } else {
+          return null;
```
that makes the method @Nullable

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableHelper.java:
```diff
@@ -68,7 +79,11 @@ static boolean hasInstanceWithRetry(final DatasetFramework datasetFramework,
+        if (!stopping) {
+          return datasetFramework.hasInstance(datasetId);
+        } else {
+          return null;
```
same here

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -100,7 +100,7 @@ private void deleteColumns(SortedMap<byte[], ? extends SortedMap<byte[], Long>>
           v2HBaseTable.delete(row, columns);
         } catch (Exception e) {
-          handleV2TableException(e);
+          handleV2TableException(e, "delete");
```
is it possible to include the table name? Like in "Delete operation for table 'metrics.v2.s.60' failed"

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java:
```diff
@@ -148,15 +148,15 @@ private Row getV2TableScannerNextRow() {
             }
           } catch (DatasetManagementException dme) {
-            LOG.error("Exception while checking dataset exists", dme);
+            LOG.error("Exception while checking v2 metrics table dataset exists", dme);
```
why not include the table name?

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -75,20 +75,36 @@ public void run() {
                MetricsTable v3MetricsTable =
                  MigrationTableHelper.getDatasetWithRetry(datasetFramework, v3MetricsDatasetId)) {
+            if (v2MetricsTable == null || v3MetricsTable == null) {
+              // we get null when we are stopping, so safe to break
```
this is ambiguous. You also get null if the table does not exist. Returning null when stopping is not the right answer.

anew commented on cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java:
```diff
@@ -75,20 +75,36 @@ public void run() {
+              break;
+            }
+            // now transfer is complete; its safe to delete the v2 metrics table
+            if (MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, v2MetricsTableId)) {
```
it may be better to throw InterruptedException from the retryable methods to indicate that your're stopping. You are already handling that exception correctly.

anew reviewed (commented):
I have just a few more comments, we are almost there :)

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedHBaseMetricsTable.java:
```diff
@@ -100,7 +100,8 @@ private void deleteColumns(SortedMap<byte[], ? extends SortedMap<byte[], Long>>
           v2HBaseTable.delete(row, columns);
         } catch (Exception e) {
-          handleV2TableException(e, "delete");
+          handleV2TableException(e, String.format("Delete operation failed for table %s",
```
you could just pass the name of the operation and the name of the table to the method...

anew reviewed (commented):
only one minor comment, up to you whether you want to change that. LGTM other than that.

gsps1 commented:
@anew updated that log change, rebased commits, have a green build here https://builds.cask.co/browse/CDAP-RUT1265-10 
merging.
